artiq/test/coredevice/test_moninj.py

```diff
@@ -66,3 +71,237 @@ def injection_status_cb(channel, override, value):
+        }
+        ftws = {}
+        notifications_out = []
+
+        loop = asyncio.new_event_loop()
+        try:
+            with self.open_comm_session(loop, notifications_out):
+                for name, urukul in self.urukuls["AD9912"].items():
+                    idx = urukul.chip_select - 4
+                    self.kernel.write_raw(urukul, target_ftws[idx])
+                    ftws[idx] = self.kernel.read_raw(urukul)[0] & (2**48 - 1)
+            final_values = {probe: value for _, probe, value in notifications_out}
+            assert final_values == ftws == target_ftws
+        finally:
+            loop.close()
+
+    def test_frequency(self):
+        target_freqs = {
+            0: 15 * MHz,
+            1: 42.5 * MHz,
+            2: 98.765 * MHz,
+            3: 128.128 * MHz
+        }
+        freqs = {}
+        notifications_out = []
+
+        loop = asyncio.new_event_loop()
+        try:
+            with self.open_comm_session(loop, notifications_out):
+                for name, urukul in self.urukuls_all().items():
+                    idx = urukul.chip_select - 4
+                    self.kernel.write(urukul, target_freqs[idx])
+                    freqs[idx] = self.kernel.read(urukul)[0]
+            for key, value in freqs.items():
+                assert numpy.isclose(value, target_freqs[key], rtol=1e-06)
+        finally:
+            loop.close()
+
+    def test_frequency_with_turns(self):
+        target_freqs = {
+            0: 15 * MHz,
+            1: 42.5 * MHz,
+            2: 98.765 * MHz,
+            3: 128.128 * MHz
+        }
+        freqs = {}
+        notifications_out = []
+
+        loop = asyncio.new_event_loop()
+        try:
+            with self.open_comm_session(loop, notifications_out):
+                for name, urukul in self.urukuls_all().items():
+                    idx = urukul.chip_select - 4
+                    self.kernel.write(urukul, target_freqs[idx], 123.4)
+                    freqs[idx] = self.kernel.read(urukul)[0]
+            for key, value in freqs.items():
+                assert numpy.isclose(value, target_freqs[key], rtol=1e-06)
+        finally:
+            loop.close()
+
+    def test_ad9910_set_freq_phase_amp(self):
+        self.ensure_ad9910_only()
+
+        target_freqs = {
+            0: 10 * MHz,
+            1: 11 * MHz,
+            2: 12 * MHz,
+            3: 13 * MHz
+        }
+        freqs = {}
+        notifications_out = []
+
+        loop = asyncio.new_event_loop()
+        try:
+            with self.open_comm_session(loop, notifications_out):
+                for name, urukul in self.urukuls["AD9910"].items():
+                    idx = urukul.chip_select - 4
+                    self.kernel.write_with_amp(urukul, target_freqs[idx], 123.4, 0.25)
+                    freqs[idx] = self.kernel.read(urukul)[0]
+            for key, value in freqs.items():
+                assert numpy.isclose(value, target_freqs[key], rtol=1e-06)
```

Comment:
Use the ``unittest`` methods not ``assert``

Reply:
done, but there's a slightly little issue on using assertAlmostEqual